File: python-client/src/crypto/authentication.py
import os
import time
import base64
import logging
import secrets
from pathlib import Path
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.exceptions import InvalidSignature

logger = logging.getLogger(__name__)

class PeerAuthentication:
    """Handles peer authentication using digital signatures"""
    
    def __init__(self, peer_id, contact_manager):
        self.peer_id = peer_id
        self.contact_manager = contact_manager
        self.storage_path = Path.home() / '.p2p-share' / 'keys'
        self.storage_path.mkdir(parents=True, exist_ok=True)
        self.private_key_path = self.storage_path / 'private.pem'
        self.public_key_path = self.storage_path / 'public.pem'
        
        # Load our own keys
        self.private_key = None
        self.public_key = None
        self.load_keys()
        
        # Track verification challenges
        self.pending_challenges = {}
        # Format: {challenge_id: {peer_id, challenge, timestamp}}
        
        logger.info(f"Authentication system initialized for peer {peer_id}")
    
    def load_keys(self):
        """Load the peer's own keys"""
        try:
            # Import the key functions only when needed
            from crypto.keys import load_private_key, load_public_key
            
            self.private_key = load_private_key(self.private_key_path)
            self.public_key = load_public_key(self.public_key_path)
            logger.info("Successfully loaded key pair")
            
            # Export public key in PEM format for sharing
            self.public_key_pem = self.public_key.public_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PublicFormat.SubjectPublicKeyInfo
            ).decode('utf-8')
            
            return True
        except Exception as e:
            logger.error(f"Failed to load keys: {e}")
            return False
    
    def get_public_key_pem(self):
        """Return our public key in PEM format for sharing"""
        return self.public_key_pem
    
    def create_challenge(self, peer_id):
        """Create a new authentication challenge for a peer"""
        challenge = secrets.token_bytes(32)  # 256-bit random challenge
        challenge_id = secrets.token_hex(8)  # Unique ID for this challenge
        challenge_b64 = base64.b64encode(challenge).decode('utf-8')
        
        # Store challenge for later verification
        self.pending_challenges[challenge_id] = {
            "peer_id": peer_id,
            "challenge": challenge,
            "timestamp": time.time()
        }
        
        # Clean up old challenges (older than 5 minutes)
        self._cleanup_challenges()
        
        return {
            "challenge_id": challenge_id,
            "challenge_b64": challenge_b64
        }
    
    def _cleanup_challenges(self):
        """Remove challenges older than 5 minutes"""
        current_time = time.time()
        expired_challenges = []
        
        for challenge_id, data in self.pending_challenges.items():
            if current_time - data["timestamp"] > 300:  # 5 minutes
                expired_challenges.append(challenge_id)
        
        for challenge_id in expired_challenges:
            del self.pending_challenges[challenge_id]
    
    def sign_challenge(self, challenge_b64):
        """Sign a challenge from another peer"""
        try:
            challenge = base64.b64decode(challenge_b64)
            
            # Create signature using our private key
            signature = self.private_key.sign(
                challenge,
                padding.PSS(
                    mgf=padding.MGF1(hashes.SHA256()),
                    salt_length=padding.PSS.MAX_LENGTH
                ),
                hashes.SHA256()
            )
            
            # Return base64 encoded signature
            return base64.b64encode(signature).decode('utf-8')
        except Exception as e:
            logger.error(f"Error signing challenge: {e}")
            return None
    
    def verify_signature(self, peer_id, challenge_id, signature_b64):
        """Verify a peer's signature of our challenge"""
        try:
            # Get the challenge
            if challenge_id not in self.pending_challenges:
                logger.error(f"Challenge {challenge_id} not found or expired")
                return False
            
            challenge_data = self.pending_challenges[challenge_id]
            if challenge_data["peer_id"] != peer_id:
                logger.error(f"Challenge was not created for peer {peer_id}")
                return False
            
            # Get the peer's public key
            contact = self.contact_manager.get_trusted_contact(peer_id)
            if not contact:
                logger.error(f"No public key found for peer {peer_id}")
                return False
            
            # Load the peer's public key
            peer_public_key = serialization.load_pem_public_key(
                contact["public_key"].encode('utf-8')
            )
            
            # Decode the signature
            signature = base64.b64decode(signature_b64)
            
            # Verify the signature
            try:
                peer_public_key.verify(
                    signature,
                    challenge_data["challenge"],
                    padding.PSS(
                        mgf=padding.MGF1(hashes.SHA256()),
                        salt_length=padding.PSS.MAX_LENGTH
                    ),
                    hashes.SHA256()
                )
                # If we get here, the signature is valid
                logger.info(f"Successfully verified signature from peer {peer_id}")
                
                # Cleanup the challenge
                del self.pending_challenges[challenge_id]
                
                # Update last seen time
                self.contact_manager.update_last_seen(peer_id)
                
                return True
            except InvalidSignature:
                logger.warning(f"Invalid signature from peer {peer_id}")
                return False
        except Exception as e:
            logger.error(f"Error verifying signature: {e}")
            return False
    
    def load_peer_public_key(self, public_key_pem):
        """Load a peer's public key from PEM format"""
        try:
            return serialization.load_pem_public_key(
                public_key_pem.encode('utf-8')
            )
        except Exception as e:
            logger.error(f"Error loading peer public key: {e}")
            return None

File: go-client/internal/crypto/authentication.go
package crypto

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// PendingChallenge represents a challenge sent to a peer
type PendingChallenge struct {
	PeerID    string
	Challenge []byte
	Timestamp float64
}

// PeerAuthentication handles authentication between peers
type PeerAuthentication struct {
	PeerID            string
	ContactManager    *ContactManager
	PrivateKey        *rsa.PrivateKey
	PublicKey         *rsa.PublicKey
	PublicKeyPEM      string
	PrivateKeyPath    string
	PublicKeyPath     string
	PendingChallenges map[string]PendingChallenge
}

// NewPeerAuthentication creates a new PeerAuthentication instance
func NewPeerAuthentication(peerID string, contactManager *ContactManager) (*PeerAuthentication, error) {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return nil, fmt.Errorf("error getting home directory: %v", err)
	}

	storagePath := filepath.Join(homeDir, ".p2p-share", "keys")
	err = os.MkdirAll(storagePath, 0755)
	if err != nil {
		return nil, fmt.Errorf("error creating keys directory: %v", err)
	}

	privateKeyPath := filepath.Join(storagePath, "private.pem")
	publicKeyPath := filepath.Join(storagePath, "public.pem")

	auth := &PeerAuthentication{
		PeerID:            peerID,
		ContactManager:    contactManager,
		PrivateKeyPath:    privateKeyPath,
		PublicKeyPath:     publicKeyPath,
		PendingChallenges: make(map[string]PendingChallenge),
	}

	// Load keys
	err = auth.LoadKeys()
	if err != nil {
		return nil, fmt.Errorf("error loading keys: %v", err)
	}

	return auth, nil
}

// LoadKeys loads the RSA key pair
func (pa *PeerAuthentication) LoadKeys() error {
	// Load private key
	privateKey, err := LoadPrivateKey(pa.PrivateKeyPath)
	if err != nil {
		return fmt.Errorf("failed to load private key: %v", err)
	}
	pa.PrivateKey = privateKey

	// Load public key
	publicKey, err := LoadPublicKey(pa.PublicKeyPath)
	if err != nil {
		return fmt.Errorf("failed to load public key: %v", err)
	}
	pa.PublicKey = publicKey

	// Export public key in PEM format for sharing
	pubKeyBytes, err := x509.MarshalPKIXPublicKey(publicKey)
	if err != nil {
		return fmt.Errorf("failed to marshal public key: %v", err)
	}

	pubKeyPEM := string(pem.EncodeToMemory(&pem.Block{
		Type:  "PUBLIC KEY",
		Bytes: pubKeyBytes,
	}))
	pa.PublicKeyPEM = pubKeyPEM

	fmt.Println("Successfully loaded key pair")
	return nil
}

// GetPublicKeyPEM returns the PEM encoded public key
func (pa *PeerAuthentication) GetPublicKeyPEM() string {
	return pa.PublicKeyPEM
}

// CreateChallenge creates a new authentication challenge for a peer
// In internal/crypto/authentication.go, modify the CreateChallenge method:

func (pa *PeerAuthentication) CreateChallenge(peerID string) (string, string, error) {
	// Generate more unique challenge ID
	challengeIDBytes := make([]byte, 16) // Increase from 8 to 16
	_, err := rand.Read(challengeIDBytes)
	if err != nil {
		return "", "", fmt.Errorf("error generating challenge ID: %v", err)
	}
	// Use timestamped ID to reduce collision chance
	timestampPrefix := fmt.Sprintf("%d_", time.Now().UnixNano())
	challengeIDStr := timestampPrefix + fmt.Sprintf("%x", challengeIDBytes)

	// Generate random 32-byte challenge
	challenge := make([]byte, 32)
	_, err = rand.Read(challenge)
	if err != nil {
		return "", "", fmt.Errorf("error generating challenge: %v", err)
	}

	// Store challenge for later verification with more data
	pa.PendingChallenges[challengeIDStr] = PendingChallenge{
		PeerID:    peerID,
		Challenge: challenge,
		Timestamp: float64(time.Now().Unix()),
	}

	// Clean up old challenges
	pa.CleanupChallenges()

	// Return base64 encoded challenge
	challengeB64 := base64.StdEncoding.EncodeToString(challenge)
	return challengeIDStr, challengeB64, nil
}

// CleanupChallenges removes challenges older than 5 minutes
func (pa *PeerAuthentication) CleanupChallenges() {
	currentTime := float64(time.Now().Unix())
	var expiredChallenges []string

	for challengeID, data := range pa.PendingChallenges {
		if currentTime-data.Timestamp > 300 { // 5 minutes
			expiredChallenges = append(expiredChallenges, challengeID)
		}
	}

	for _, challengeID := range expiredChallenges {
		delete(pa.PendingChallenges, challengeID)
	}
}

// SignChallenge signs a challenge from another peer
func (pa *PeerAuthentication) SignChallenge(challengeB64 string) (string, error) {
	// Decode the challenge
	challenge, err := base64.StdEncoding.DecodeString(challengeB64)
	if err != nil {
		return "", fmt.Errorf("error decoding challenge: %v", err)
	}

	// Hash the challenge
	hashed := sha256.Sum256(challenge)

	// Sign the hash
	signature, err := rsa.SignPSS(rand.Reader, pa.PrivateKey, crypto.SHA256, hashed[:], &rsa.PSSOptions{
		SaltLength: rsa.PSSSaltLengthAuto,
	})
	if err != nil {
		return "", fmt.Errorf("error signing challenge: %v", err)
	}

	// Return base64 encoded signature
	signatureB64 := base64.StdEncoding.EncodeToString(signature)
	return signatureB64, nil
}

// VerifySignature verifies a peer's signature of our challenge
func (pa *PeerAuthentication) VerifySignature(peerID, challengeID, signatureB64 string, publicKeyPEM string) (bool, error) {
	// Get the challenge
	challengeData, exists := pa.PendingChallenges[challengeID]
	if !exists {
		return false, fmt.Errorf("challenge %s not found or expired", challengeID)
	}

	if challengeData.PeerID != peerID {
		return false, fmt.Errorf("challenge was not created for peer %s", peerID)
	}

	var block *pem.Block
	var pubInterface interface{}
	var err error

	// If a public key PEM is provided directly, use it
	if publicKeyPEM != "" {
		block, _ = pem.Decode([]byte(publicKeyPEM))
		if block == nil || block.Type != "PUBLIC KEY" {
			return false, fmt.Errorf("failed to decode PEM block containing public key")
		}

		pubInterface, err = x509.ParsePKIXPublicKey(block.Bytes)
		if err != nil {
			return false, fmt.Errorf("failed to parse public key: %v", err)
		}

		pubKey, ok := pubInterface.(*rsa.PublicKey)
		if !ok {
			return false, fmt.Errorf("not an RSA public key")
		}

		// Decode the signature
		signature, err := base64.StdEncoding.DecodeString(signatureB64)
		if err != nil {
			return false, fmt.Errorf("error decoding signature: %v", err)
		}

		// Hash the challenge
		hashed := sha256.Sum256(challengeData.Challenge)

		// Verify the signature
		err = rsa.VerifyPSS(pubKey, crypto.SHA256, hashed[:], signature, &rsa.PSSOptions{
			SaltLength: rsa.PSSSaltLengthAuto,
		})
		if err != nil {
			return false, fmt.Errorf("invalid signature: %v", err)
		}
	} else {
		// Otherwise, try to get the public key from the contact manager
		contact, exists := pa.ContactManager.GetTrustedContact(peerID)
		if !exists {
			return false, fmt.Errorf("no public key found for peer %s", peerID)
		}

		// Load the peer's public key
		block, _ = pem.Decode([]byte(contact.PublicKey))
		if block == nil || block.Type != "PUBLIC KEY" {
			return false, fmt.Errorf("failed to decode PEM block containing public key")
		}

		pubInterface, err = x509.ParsePKIXPublicKey(block.Bytes)
		if err != nil {
			return false, fmt.Errorf("failed to parse public key: %v", err)
		}

		pubKey, ok := pubInterface.(*rsa.PublicKey)
		if !ok {
			return false, fmt.Errorf("not an RSA public key")
		}

		// Decode the signature
		signature, err := base64.StdEncoding.DecodeString(signatureB64)
		if err != nil {
			return false, fmt.Errorf("error decoding signature: %v", err)
		}

		// Hash the challenge
		hashed := sha256.Sum256(challengeData.Challenge)

		// Verify the signature
		err = rsa.VerifyPSS(pubKey, crypto.SHA256, hashed[:], signature, &rsa.PSSOptions{
			SaltLength: rsa.PSSSaltLengthAuto,
		})
		if err != nil {
			return false, fmt.Errorf("invalid signature: %v", err)
		}
	}

	// Signature is valid
	fmt.Printf("Successfully verified signature from peer %s\n", peerID)

	// Clean up the challenge
	delete(pa.PendingChallenges, challengeID)

	// Update last seen time
	pa.ContactManager.UpdateLastSeen(peerID)

	return true, nil
}

// LoadPeerPublicKey loads and parses a peer's public key from PEM format
func (pa *PeerAuthentication) LoadPeerPublicKey(publicKeyPEM string) (*rsa.PublicKey, error) {
	block, _ := pem.Decode([]byte(publicKeyPEM))
	if block == nil || block.Type != "PUBLIC KEY" {
		return nil, fmt.Errorf("failed to decode PEM block containing public key")
	}

	pubInterface, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("failed to parse public key: %v", err)
	}

	pubKey, ok := pubInterface.(*rsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("not an RSA public key")
	}

	return pubKey, nil
}

// Add this function to a suitable file in the crypto package, e.g., authentication.go
func GetPeerID() string {
	if authentication == nil {
		return ""
	}
	return authentication.PeerID
}
